lib/pos_blockchain.go

```diff
@@ -0,0 +1,1134 @@
+package lib
+
+import (
+	"math"
+	"time"
+
+	"github.com/deso-protocol/core/collections"
+	"github.com/deso-protocol/core/consensus"
+	"github.com/dgraph-io/badger/v3"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+)
+
+// processBlockPoS runs the Fast-Hotstuff block connect and commit rule as follows:
+//  1. Determine if we're missing the parent block of this block.
+//     If so, return the hash of the missing block and add this block to the orphans list.
+//  2. Validate the incoming block, its header, its block height, the leader, and its QCs (vote or timeout)
+//  3. Store the block in the block index and save to DB.
+//  4. try to apply the incoming block as the tip (performing reorgs as necessary). If it can't be applied, we exit here.
+//  5. Run the commit rule - If applicable, flushes the incoming block's grandparent to the DB
+func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, verifySignatures bool) (_success bool, _isOrphan bool, _missingBlockHashes []*BlockHash, _err error) {
+	// If we can't hash the block, we can never store in the block index and we should throw it out immediately.
+	if _, err := block.Hash(); err != nil {
+		return false, false, nil, errors.New("processBlockPoS: Problem hashing block")
+	}
+	// Get all the blocks between the current block and the committed tip. If the block
+	// is an orphan, then we store it after performing basic validations.
+	// If the block extends from any committed block other than the committed tip,
+	// then we throw it away.
+	lineageFromCommittedTip, err := bc.getLineageFromCommittedTip(block)
```

Comment:
How is this not the same as the following block?
```
	lineageFromCommittedTip, err := bc.getLineageFromCommittedTip(block)
	if err == RuleErrorMissingAncestorBlock {
		// In this case, the block is an orphan that does not extend from any blocks
		// on our best chain. Try to process the orphan by running basic validations.
		// If it passes basic integrity checks, we'll store it with the hope that we
		// will eventually get a parent that connects to our best chain.
		missingBlockHashes := []*BlockHash{block.Header.PrevBlockHash}
		// FIXME: I sketeched some of these steps but not all...
		// Step 0: I think we still need to do some basic validation on this block, like
		// verifying that it's signed by the leader for example to prevent spamming.
		// I didn't do that.
		// Step 1: Create a new BlockNode for this block with status STORED.
		// Step 2: Add it to the blockIndexByHash and store it in Badger. This is handled by addBlockToBlockIndex
		return false, true, missingBlockHashes, bc.processOrphanBlockPoS(block)
	}
	if err != nil {
		return false, false, nil, errors.Wrap(err, "processBlockPoS: Problem getting lineage from committed tip: UNEXPECTED ERROR!!!")
	}
```

Reply:
Because I forgot to add a call to `storeValidateFailedBlockInBlockIndex`!! Good catch. If we hit any of the specified errors, we know that we'll never be able to validate this block, so we can go ahead and mark it as ValidateFailed.